### DISCUSSION TITLE
Possible issue with color::from_hsl

### DISCUSSION THREAD

**Discussion by ErrorFlynn** [cnjinhao/nana#594] [Q&A] [answered]
The method `color::from_hsl` seems to produce incorrect results. For example, `color.from_hsl(210, .68, .80);` produces `{r_=169.32000000000005 g_=204.00000000000000 b_=238.67999999999998 ...}`. The correct RGB values are `171, 205, 239` (`#abcdef`), as any online color picker will tell you (for example: https://colorpicker.me/#abcdef).

The reason I said "possible issue", is because I wrote my own conversion function, and it produces the same results as `from_hsl`! Are both my code and `from_hsl` using the wrong formula, or am I missing something? I'm guessing something else is happening, but can't imagine what. Here's my code:

````
color hsl2rgb(double h, double s, double l)
{
	double r, g, b;
	if(s == 0.0)
	{
		r = l * 255.0;
		g = r;
		b = r;
	}
	else
	{
		double t1;
		if(l < .5)
			t1 = l * (1.0 + s);
		else t1 = l + s - l * s;
		double t2 {2 * l - t1};
		h /= 360;

		auto h2rgb = [&](double val)
		{
			if(val < 0.0) val += 1.0;
			else if(val > 1.0) val -= 1.0;
			if(6.0 * val < 1.0)
				return t2 + (t1 - t2) * 6 * val;
			if(2 * val < 1.0)
				return t1;
			if(3 * val < 2.0)
				return t2 + (t1 - t2) * (.666 - val) * 6;
			return t2;
		};

		r = h2rgb(h + .333);
		g = h2rgb(h);
		b = h2rgb(h - .333);
	}
	return {static_cast<unsigned>(round(r*255)), static_cast<unsigned>(round(g*255)), static_cast<unsigned>(round(b*255))};
}
````

**Comment by ashbob999** [comment, marked as answer]
After doing some searching and implementing the same algorithm from different websites, i have gotten the exact same value each time for `hsl(210, 0.68, 0.80)` produces `r=169, g=204, b=239` (`#A9CCEF`). From this website: [https://convertacolor.com/](https://convertacolor.com/).

There does not seem to be a problem with the algorithm but I think the problem lies in the colour picker websites ([partially explained here](https://stackoverflow.com/questions/57804274/css-why-does-converting-a-hex-color-to-hsl-and-then-back-to-hex-lead-to-a-diffe))

**Example**
`#ABCDEF` -> `hsl(210, 68%, 80.4%)` they round 80.4% to 80%
So we get `hsl(210, 0.68, 0.80)` -> '#A9CCEF'
So if the websites round the `saturation/lightness` to an integer (if percentage) or 2 decimal places (if double), they will then be giving incorrect conversion values.

**Reply by ErrorFlynn** [reply]
Wow, I can't believe the websites are wrong, and I can't believe I got tripped up by this. Thanks a lot for the explanation!
